### PR TITLE
1 new exercise: protein-translation

### PR DESCRIPTION
**Pull request by kapitaali**
Adding 1 new exercise to COBOL track: protein-translation

Based on last pull request attempt, all GOBACKs have been removed from code.

### CODE REVIEW

**Comment by kapitaali** [comment]
Forgot config.json once again, but it has been fixed in my branch.

**Comment by axtens** [comment]
I've sent a pile of comments via the forum. See https://forum.exercism.org/t/requirements-for-cut-and-proof-ci-cob-files/10818

File: exercises/practice/protein-translation/bin/fetch-cobolcheck.ps1 [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    $latestUrl = "https://api.github.com/repos/0xE282B0/cobol-check/releases/latest"
+    Invoke-RestMethod -Uri $latestUrl -PreserveAuthorizationOnRedirect @requestOpts
+    | Select-Object -ExpandProperty assets
+    | Where-Object { $_.browser_download_url -match $FileName }
+    | Select-Object -ExpandProperty browser_download_url
+}
+
+$downloadUrl = Get-DownloadUrl
+$outputFile = Join-Path -Path $PSScriptRoot -ChildPath "cobolcheck.exe"
+Invoke-WebRequest -Uri $downloadUrl -OutFile $outputFile @requestOpts
```
Comment (kotp) [inline exercises/practice/protein-translation/bin/fetch-cobolcheck.ps1:28]:
```suggestion
Invoke-WebRequest -Uri $downloadUrl -OutFile $outputFile @requestOpts

```

File: exercises/practice/protein-translation/src/protein-translation.cob [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+       WORKING-STORAGE SECTION.
+       01 WS-INPUT                  PIC X(60).
+       01 WS-PROTEIN                PIC X(120).
+       01 WS-ERROR                  PIC X(60).
+
+       PROCEDURE DIVISION.
+
+       TRANSLATE-CODON.
+
+       TRANSLATE-RNA.
```
Comment (kotp) [inline exercises/practice/protein-translation/src/protein-translation.cob:14]:
```suggestion
       TRANSLATE-RNA.
       
```

**Review by kotp — APPROVED** [review]
The suggested changes are to make lines consistent throughout.

**Comment by axtens** [comment]
@ErikSchierboom still have the appveyor issue `COBOL / Test (Windows) / ci (pull_request)`. A separate issue in `COBOL / Test / ci (pull_request)`

**Comment by axtens** [comment]
We could stand to remove some of the `DISPLAY` statements in the various `proof.ci.cob` files as they clutter up the test runs

**Comment by axtens** [comment]
@kapitaali tests are failing ... but you knew that

**Comment by ErikSchierboom** [comment]
> Checking protein-translation practice exercise...
Cobolcheck not found, try to fetch it.
CobolCheck: INFO 2024-04-18T03:35:18.670116Z INF001: Attempting to load config from config.properties.
CobolCheck: INFO 2024-04-18T03:35:18.670267Z INF002: Loaded config successfully from config.properties.
CobolCheck: INFO 2024-04-18T03:35:18.670289Z INF003: Cobol-Check starting
CobolCheck: INFO 2024-04-18T03:35:18.670298Z INF005: Log level is INFO.
CobolCheck: INFO 2024-04-18T03:35:18.670306Z INF006: Configuration settings: production.
CobolCheck: INFO 2024-04-18T03:35:18.670637Z INF014: Error log for the test suite parser is set to: /home/runner/work/cobol/cobol/exercises/practice/protein-translation/ParserErrorLog.txt.
CobolCheck: INFO 2024-04-18T03:35:18.670897Z INF013: Output for generated COBOL test program is set to: /home/runner/work/cobol/cobol/exercises/practice/protein-translation/test.cob.
CobolCheck: INFO 2024-04-18T03:35:18.678179Z INF012: Successfully generated COBOL test program for src/protein-translation.
CobolCheck: INFO 2024-04-18T03:35:18.678224Z INF004: Cobol-Check terminating: 0
COMPILE AND RUN TEST
test.cob: in paragraph 'UT-AFTER-EACH':
test.cob:791: error: syntax error, unexpected PERFORM
protein-translation: proof solution did not pass the tests

That makes it appear like there is a syntax error somewhere in protein-translation.

> @ErikSchierboom still have the appveyor issue COBOL / Test (Windows) / ci (pull_request)

Feel free to try and fix that, I don't have much time for that.

**Comment by axtens** [comment]
> Feel free to try and fix that, I don't have much time for that.
Righto. I did get feedback from @gitmensch about that. See https://github.com/exercism/cobol/pull/126#issuecomment-2061191642 but I'll see what I can do about it in the meantime

**Comment by axtens** [comment]
The compiler error is being caused by `cobolcheck` not properly parsing 
```
AFTER-EACH
	MOVE SPACES TO WS-PROTEIN
END-AFTER
```
therefore, remove that from the CUT file and explicitly INITIALIZE WS-PROTEIN at the top of the relevant paragraphs.

**Comment by axtens** [comment]
This passes all the tests.
```
       IDENTIFICATION DIVISION.
       PROGRAM-ID. PROTEIN-TRANSLATION.
       AUTHOR. kapitaali.
       ENVIRONMENT DIVISION.
       DATA DIVISION.
       WORKING-STORAGE SECTION.
       01 WS-INPUT                  PIC X(60).
       01 WS-PROTEIN                PIC X(120).
       01 WS-ERROR                  PIC X(60). 
       01 WS-CODON                  PIC XXX.   
       01 WS-AMINOACID              PIC X(14).
       01 SLICES.
         03 SLICE PIC XXX OCCURS 20 TIMES 
            INDEXED BY SLICED.

       PROCEDURE DIVISION.
       TRANSLATE-CODON.
           INITIALIZE WS-PROTEIN WS-ERROR WS-CODON WS-AMINOACID.
           MOVE WS-INPUT TO SLICES.
           SET SLICED TO 1.
           PERFORM WITH TEST BEFORE UNTIL SLICE(SLICED) = SPACE
               MOVE SLICE(SLICED) TO WS-CODON
               PERFORM CHECK-CODON
               SET SLICED UP BY  1
           END-PERFORM.
           IF WS-AMINOACID = "ERROR" 
             MOVE "Invalid codon" TO WS-ERROR
           ELSE 
             MOVE WS-AMINOACID TO WS-PROTEIN.
           EXIT PARAGRAPH.
           
       CHECK-CODON.
           MOVE SPACES TO WS-AMINOACID.
           EVALUATE WS-CODON 
              WHEN SPACE MOVE SPACE TO WS-AMINOACID
              WHEN "AUG" MOVE "Methionine" TO WS-AMINOACID
              WHEN "UUU" MOVE "Phenylalanine" TO WS-AMINOACID
              WHEN "UUC" MOVE "Phenylalanine" TO WS-AMINOACID
              WHEN "UUA" MOVE "Leucine" TO WS-AMINOACID
              WHEN "UUG" MOVE "Leucine" TO WS-AMINOACID
              WHEN "UCU" MOVE "Serine" TO WS-AMINOACID
              WHEN "UCC" MOVE "Serine" TO WS-AMINOACID
              WHEN "UCA" MOVE "Serine" TO WS-AMINOACID
              WHEN "UCG" MOVE "Serine" TO WS-AMINOACID
              WHEN "UAU" MOVE "Tyrosine" TO WS-AMINOACID
              WHEN "UAC" MOVE "Tyrosine" TO WS-AMINOACID
              WHEN "UGU" MOVE "Cysteine" TO WS-AMINOACID
              WHEN "UGC" MOVE "Cysteine" TO WS-AMINOACID
              WHEN "UGG" MOVE "Tryptophan" TO WS-AMINOACID
              WHEN "UAA" MOVE "STOP" TO WS-AMINOACID
              WHEN "UAG" MOVE "STOP" TO WS-AMINOACID
              WHEN "UGA" MOVE "STOP" TO WS-AMINOACID
              WHEN OTHER MOVE "ERROR" TO WS-AMINOACID
           END-EVALUATE.
           EXIT PARAGRAPH.

       TRANSLATE-RNA.
           INITIALIZE WS-PROTEIN WS-ERROR WS-CODON WS-AMINOACID.
           MOVE WS-INPUT TO SLICES.
           SET SLICED TO 1.
           PERFORM WITH TEST AFTER UNTIL SLICE(SLICED) = SPACE
               MOVE SLICE(SLICED) TO WS-CODON
               PERFORM CHECK-CODON
               IF WS-AMINOACID = "STOP"
                 EXIT PERFORM
               END-IF
               IF WS-AMINOACID = "ERROR"
                 MOVE ",ERROR" TO WS-PROTEIN
                 EXIT PERFORM
               END-IF
               STRING WS-PROTEIN DELIMITED BY SPACE 
                 "," DELIMITED BY SIZE
                 WS-AMINOACID DELIMITED BY SPACE 
                 INTO WS-PROTEIN
               END-STRING
               SET SLICED UP BY 1
           END-PERFORM.
           MOVE WS-PROTEIN(2:) TO WS-PROTEIN.
           IF WS-PROTEIN = "ERROR"
             MOVE "Invalid codon" TO WS-ERROR.
           EXIT PARAGRAPH.
```

**Comment by GitMensch** [comment]
If this is _also_ about good COBOL, I'd recommend to drop the `EXIT PARAGRAPH` - that is seldom used and at the end of the paragraph makes no sense at all.

The complete "fall through" would commonly be seen as bad style. It may should be adjusted to instead of 

```cobol
       TRANSLATE-CODON.
           INITIALIZE WS-PROTEIN WS-ERROR WS-CODON WS-AMINOACID.
           MOVE WS-INPUT TO SLICES.
           SET SLICED TO 1.
           PERFORM WITH TEST BEFORE UNTIL SLICE(SLICED) = SPACE
               MOVE SLICE(SLICED) TO WS-CODON
               PERFORM CHECK-CODON
               SET SLICED UP BY  1
           END-PERFORM.
           IF WS-AMINOACID = "ERROR" 
             MOVE "Invalid codon" TO WS-ERROR
           ELSE 
             MOVE WS-AMINOACID TO WS-PROTEIN.
           EXIT PARAGRAPH.
```

be:


```cobol
       TRANSLATE-MAIN.
           INITIALIZE WS-PROTEIN WS-ERROR WS-CODON WS-AMINOACID.
           MOVE WS-INPUT TO SLICES.

           PERFORM TRANSLATE-CODON.
           IF WS-AMINOACID = "ERROR" 
             MOVE "Invalid codon" TO WS-ERROR
             GOBACK.

           MOVE WS-AMINOACID TO WS-PROTEIN.
           PERFORM TRANSLATE-RNA.

           GOBACK.

       TRANSLATE-CODON.
           SET SLICED TO 1.
           PERFORM WITH TEST BEFORE UNTIL SLICE(SLICED) = SPACE
               MOVE SLICE(SLICED) TO WS-CODON
               PERFORM CHECK-CODON
               SET SLICED UP BY  1
           END-PERFORM.
```

**Comment by kapitaali** [comment]
> If this is _also_ about good COBOL, I'd recommend to drop the `EXIT PARAGRAPH` - that is seldom used and at the end of the paragraph makes no sense at all.

it wasn't about good COBOL last time I submitted exercises, the only requirement was that the code passes the tests

GOBACKs supposedly messed up something with testing, so I removed them and used EXIT PARAGRAPH

**Comment by ErikSchierboom** [comment]
> it wasn't about good COBOL last time I submitted exercises, the only requirement was that the code passes the tests

That is still 100% the case.

**Comment by GitMensch** [comment]
> GOBACKs supposedly messed up something with testing, so I removed them and used `EXIT PARAGRAPH`

`GOBACK` gets you to the caller (which may be the operating system), you'd only use one if you don't want to "fall out of the program", either in the current fall-through state at the very end, or with a "main" paragraph `PERFORM`ing the other paragraphs at its end.

As noted: `EXIT PARAGRAPH` in this code doesn't do anything here, because it gets you directly before the next paragraph - which is the exact place where this statement is.

"The rest is a style question and in COBOL those are commonly set fixed in the team/working place (you will find a bunch of those heavily disagreeing; and there may be one where the fall-through is preferred [I've so far just met styles where this was allowed within sections, but not for the whole program]).

**Comment by axtens** [comment]
We've found that GOBACK and STOP RUN confuse COBOLCHECK

**Comment by GitMensch** [comment]
I'm not sure what you reference with COBOLCHECK, but it seems that part is buggy and may should be improved?

**Comment by axtens** [comment]
FYI https://openmainframeproject.org/projects/cobol-check/ 
github:  https://github.com/openmainframeproject/cobol-check/

**Comment by axtens** [comment]
As asked in the forum, How do I rebase https://github.com/exercism/cobol/pull/125 so that it uses the new setup? I thought this was what was needed
```
bugmagnet@bugmagnet-wasta:~/Projects/Everything_Exercism/exercism-tracks/cobol$ git pull upstream main
remote: Enumerating objects: 1, done.
remote: Counting objects: 100% (1/1), done.
remote: Total 1 (delta 0), reused 0 (delta 0), pack-reused 0
Unpacking objects: 100% (1/1), 927 bytes | 927.00 KiB/s, done.
From https://github.com/exercism/cobol
 * branch            main       -> FETCH_HEAD
   be1cc08..40ed166  main       -> upstream/main
Successfully rebased and updated refs/heads/protein-translation.
```

**Comment by axtens** [comment]
@kotp all the CI passes. Lots of hard work done by @ErikSchierboom

**Review by kotp — APPROVED** [review]

**Comment by axtens** [comment]
@ErikSchierboom why can't I merge this?

**Comment by ErikSchierboom** [comment]
A maintainer has to approve it first

**Review by axtens — APPROVED** [review]

**Pull request merged**